src/CSVMsg.chpl

```diff
@@ -197,7 +197,32 @@ module CSVMsg {
             }
         }
 
-        coforall loc in Locales do on loc{
+        // only use 90% of getMemLimit to give ourselves some buffer
+        // these are each per locale measurements
+        const memLim = (getMemLimit() * (.9)):uint;
+        const memUsed = getMemUsed();
+
+        // calculate the amount of memory required to write the csv file all in one go
+        var memRequired: uint = 0;
+        var numStrings = 0;
+        for (cname, dt) in zip(dtypes, datasets) {
+            if dt == "Strings" {
+                var ss = new SegString("", toSegStringSymEntry(st.lookup(cname)));
+                // strings vals are uint(8), so they're only one byte but overcount
+                // because they're not evenly distributed across locales
+                memRequired += (ss.values.a.size * 1.5):int;
+                numStrings += 1;
+            }
+        }
+        // assume pdarrays are 64 bits / 8 bytes (this is an overestimate for bools, but that's fine)
+        memRequired += ((ndsets-numStrings) * row_count * 8);
+        memRequired = (memRequired:real / numLocales):uint;
+
+        // if we can put the entire df in mem (i.e. memLim > (memUsed + memRequired)), then we can do it all in one go.
+        // Otherwise we can only put (memLim - memUsed) in memory at a time, so that's our batchSize
+        const numBatches = if memLim > (memUsed + memRequired) then 1 else ceil(memRequired:real / (memLim - memUsed)):int;
```

Comment:
Also, what happens if `memUsed > memLim`?

Reply:
yaaaah that would give us a negative number of batches, which is def not what we want. It likely means we'll run out of memory since we'd already be at 90% of our memory limit. I think we should prob check for that.

If we've exceeded 90% of mem on the system then we will break the chunks into at most size 5% to give us some buffer

Reply:
If we were over 95% and trying to write a large csv, we will likely run out of memory during the operation, but I feel like that's reasonable

Reply:
so we can use this new and improved (hopefully actually useful) ternary! lol
```chapel
const numBatches = if memLim > memUsed then ceil(memRequired:real / (memLim - memUsed)):int else ceil(memRequired:real / (memLim / 18)):int;
```